In [4]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [5]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [6]:
# import excel file November2012.xlsx
tn_2012=pd.read_excel('../data/2012_election/November2012.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','BName1','Tally1','BName2','Tally2',
                               'BName3','Tally3','BName4','Tally4','BName5','Tally5','BName6','Tally6','BName7','Tally7',
                               'BName8','Tally8','BName9','Tally9','BName10','Tally10'])
tn_2012

COUNTY       PRECINCT                                   OFFICENAME  \
0      ANDERSON  Andersonville                      United States President   
1      ANDERSON  Andersonville                         United States Senate   
2      ANDERSON  Andersonville                         United States Senate   
3      ANDERSON  Andersonville                         United States Senate   
4      ANDERSON  Andersonville       U.S. House of Representatives District   
...         ...            ...                                          ...   
14087    WILSON    Provisional                         United States Senate   
14088    WILSON    Provisional                         United States Senate   
14089    WILSON    Provisional       U.S. House of Representatives District   
14090    WILSON    Provisional  Tennessee House of Representatives District   
14091    WILSON    Provisional  Tennessee House of Representatives District   

              ELECTDATE                       BName1  Tally1  \
0      November 6, 2012          . Mitt Romney - (R)    1110   
1      November 6, 2012           . Bob Corker - (R)    1067   
2      November 6, 2012      . Write-In - Larry Crim       0   
3      November 6, 2012  . Write-In - Mary D. Wright       0   
4      November 6, 2012    . Chuck Fleischmann - (R)     996   
...                 ...                          ...     ...   
14087  November 6, 2012      . Write-In - Larry Crim       0   
14088  November 6, 2012  . Write-In - Mary D. Wright       0   
14089  November 6, 2012          . Diane Black - (R)      36   
14090  November 6, 2012         . Mark A. Pody - (R)      16   
14091  November 6, 2012           . Susan Lynn - (R)      30   

                             BName2  Tally2                       BName3  \
0              . Barack Obama - (D)     267         . Virgil Goode - (C)   
1           . Mark E. Clayton - (D)     216         . Kermit Steck - (C)   
2      . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
3                               NaN       0                          NaN   
4          . Mary M. Headrick - (D)     282     . Matthew Deniston - (I)   
...                             ...     ...                          ...   
14087  . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
14088                           NaN       0                          NaN   
14089             . Pat Riley - (G)       9        . Scott Beasley - (I)   
14090                           NaN       0                          NaN   
14091                           NaN       0                          NaN   

       Tally3                           BName4  Tally4  \
0           3               . Jill Stein - (G)       0   
1          17          . Martin Pleasant - (G)      11   
2           0   . Write-In - De Wayne A. Jones       0   
3           0                              NaN       0   
4          48        . Write-In - Richard Hall       0   
...       ...                              ...     ...   
14087       0   . Write-In - De Wayne A. Jones       0   
14088       0                              NaN       0   
14089       9  . Write-In - Mary Amanda Keifer       0   
14090       0                              NaN       0   
14091       0                              NaN       0   

                                     BName5  Tally5  \
0          . Ross C. "Rocky" Anderson - (I)       0   
1                  . Shaun E. Crowell - (I)      18   
2                 . Write-In - Jacob Maurer       0   
3                                       NaN       0   
4      . Write-In - Lady J Jean Howard-Hill       0   
...                                     ...     ...   
14087             . Write-In - Jacob Maurer       0   
14088                                   NaN       0   
14089          . Write-In - Rachel Robinson       0   
14090                                   NaN       0   
14091                                   NaN       0   

                    

In [8]:
tn_2012.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14092 entries, 0 to 14091
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COUNTY      14092 non-null  object
 1   PRECINCT    14092 non-null  object
 2   OFFICENAME  14092 non-null  object
 3   ELECTDATE   14092 non-null  object
 4   BName1      14092 non-null  object
 5   Tally1      14092 non-null  int64 
 6   BName2      10660 non-null  object
 7   Tally2      14092 non-null  int64 
 8   BName3      8449 non-null   object
 9   Tally3      14092 non-null  int64 
 10  BName4      8207 non-null   object
 11  Tally4      14092 non-null  int64 
 12  BName5      7762 non-null   object
 13  Tally5      14092 non-null  int64 
 14  BName6      7059 non-null   object
 15  Tally6      14092 non-null  int64 
 16  BName7      6489 non-null   object
 17  Tally7      14092 non-null  int64 
 18  BName8      4210 non-null   object
 19  Tally8      14092 non-null  int64 
 20  BName9

In [9]:
# merge entries in all RNAME columns into new column "name"
tn_2012_namemelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['BName1','BName2','BName3','BName4','BName5','BName6','BName7','BName8','BName9','BName10'],value_name="candidate")
# drop column "variable" created by .melt
tn_2012_namemelt=tn_2012_namemelt.drop(['variable'],axis=1)                          
tn_2012_namemelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE                    candidate  
0  November 6, 2012          . Mitt Romney - (R)  
1  November 6, 2012           . Bob Corker - (R)  
2  November 6, 2012      . Write-In - Larry Crim  
3  November 6, 2012  . Write-In - Mary D. Wright  
4  November 6, 2012    . Chuck Fleischmann - (R)

In [10]:
# merge entries in all TALLY columns into new column "vote_count"
tn_2012_pvtallymelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['Tally1','Tally2','Tally3','Tally4','Tally5','Tally6','Tally7','Tally8','Tally9','Tally10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2012_pvtallymelt=tn_2012_pvtallymelt.drop(['variable'],axis=1)
tn_2012_pvtallymelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE  vote_count  
0  November 6, 2012        1110  
1  November 6, 2012        1067  
2  November 6, 2012           0  
3  November 6, 2012           0  
4  November 6, 2012         996

In [11]:
# add "vote_count" to tn_2016_namemelt
tally=tn_2012_pvtallymelt['vote_count']
tn_2012_namemelt=tn_2012_namemelt.join(tally)
tn_2012_namemelt.head(3)

COUNTY       PRECINCT               OFFICENAME         ELECTDATE  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   
2  ANDERSON  Andersonville     United States Senate  November 6, 2012   

                 candidate  vote_count  
0      . Mitt Romney - (R)        1110  
1       . Bob Corker - (R)        1067  
2  . Write-In - Larry Crim           0

In [12]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   COUNTY      140920 non-null  object
 1   PRECINCT    140920 non-null  object
 2   OFFICENAME  140920 non-null  object
 3   ELECTDATE   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


In [13]:
# lower case column names
tn_2012_namemelt.columns=tn_2012_namemelt.columns.str.lower()
tn_2012_namemelt.head(2)

county       precinct               officename         electdate  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   

             candidate  vote_count  
0  . Mitt Romney - (R)        1110  
1   . Bob Corker - (R)        1067

In [14]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      140920 non-null  object
 1   precinct    140920 non-null  object
 2   officename  140920 non-null  object
 3   electdate   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


#### 75348 non-null names  out of 140920 enties. Confirm all are 0 votes and drop rows with no candidate listed

In [15]:
# Create dataframe "blank_candidate_2012" of null valuea in "candidate" of tn_2012_namemelt
blank_candidate_2012=tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()]
blank_candidate_2012

county       precinct                                   officename  \
14095   ANDERSON  Andersonville                         United States Senate   
14101   ANDERSON     Briceville                         United States Senate   
14107   ANDERSON       Bull Run                         United States Senate   
14113   ANDERSON        Clinton                         United States Senate   
14119   ANDERSON   Clinton High                         United States Senate   
...          ...            ...                                          ...   
140913    WILSON    Provisional                      United States President   
140916    WILSON    Provisional                         United States Senate   
140917    WILSON    Provisional       U.S. House of Representatives District   
140918    WILSON    Provisional  Tennessee House of Representatives District   
140919    WILSON    Provisional  Tennessee House of Representatives District   

               electdate candidate  vote_count  
14095   November 6, 2012       NaN           0  
14101   November 6, 2012       NaN           0  
14107   November 6, 2012       NaN           0  
14113   November 6, 2012       NaN           0  
14119   November 6, 2012       NaN           0  
...                  ...       ...         ...  
140913  November 6, 2012       NaN           0  
140916  November 6, 2012       NaN           0  
140917  November 6, 2012       NaN           0  
140918  November 6, 2012       NaN           0  
140919  November 6, 2012       NaN           0  

[65572 rows x 6 columns]

In [16]:
# confirm all are 0 counts
blank_candidate_2012['vote_count'].value_counts()

0    65572
Name: vote_count, dtype: int64

In [17]:
# drop all null candidates
tn_2012_namemelt.dropna().reset_index(drop=True)

county       precinct                              officename  \
0      ANDERSON  Andersonville                 United States President   
1      ANDERSON  Andersonville                    United States Senate   
2      ANDERSON  Andersonville                    United States Senate   
3      ANDERSON  Andersonville                    United States Senate   
4      ANDERSON  Andersonville  U.S. House of Representatives District   
...         ...            ...                                     ...   
75343    WILSON           25-2                    United States Senate   
75344    WILSON       Absentee                    United States Senate   
75345    WILSON       Absentee                    United States Senate   
75346    WILSON    Provisional                    United States Senate   
75347    WILSON    Provisional                    United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012        . Write-In - Larry Crim           0  
3      November 6, 2012    . Write-In - Mary D. Wright           0  
4      November 6, 2012      . Chuck Fleischmann - (R)         996  
...                 ...                            ...         ...  
75343  November 6, 2012  . Write-In - Angelia Stinnett           0  
75344  November 6, 2012  . Write-In - Fred R. Anderson           0  
75345  November 6, 2012  . Write-In - Angelia Stinnett           0  
75346  November 6, 2012  . Write-In - Fred R. Anderson           0  
75347  November 6, 2012  . Write-In - Angelia Stinnett           0  

[75348 rows x 6 columns]

In [19]:
# confirn all null candidates dropped
tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65572 entries, 14095 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      65572 non-null  object
 1   precinct    65572 non-null  object
 2   officename  65572 non-null  object
 3   electdate   65572 non-null  object
 4   candidate   0 non-null      object
 5   vote_count  65572 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 3.5+ MB


In [20]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      140920 non-null  object
 1   precinct    140920 non-null  object
 2   officename  140920 non-null  object
 3   electdate   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


### Look at cadidates with 0 votes, drop

In [21]:
tn_2012_namemelt_0votes=tn_2012_namemelt[tn_2012_namemelt['vote_count'] == 0]
tn_2012_namemelt_0votes

county       precinct                                   officename  \
2       ANDERSON  Andersonville                         United States Senate   
3       ANDERSON  Andersonville                         United States Senate   
8       ANDERSON     Briceville                         United States Senate   
9       ANDERSON     Briceville                         United States Senate   
14      ANDERSON       Bull Run                         United States Senate   
...          ...            ...                                          ...   
140915    WILSON    Provisional                         United States Senate   
140916    WILSON    Provisional                         United States Senate   
140917    WILSON    Provisional       U.S. House of Representatives District   
140918    WILSON    Provisional  Tennessee House of Representatives District   
140919    WILSON    Provisional  Tennessee House of Representatives District   

               electdate                      candidate  vote_count  
2       November 6, 2012        . Write-In - Larry Crim           0  
3       November 6, 2012    . Write-In - Mary D. Wright           0  
8       November 6, 2012        . Write-In - Larry Crim           0  
9       November 6, 2012    . Write-In - Mary D. Wright           0  
14      November 6, 2012        . Write-In - Larry Crim           0  
...                  ...                            ...         ...  
140915  November 6, 2012  . Write-In - Angelia Stinnett           0  
140916  November 6, 2012                            NaN           0  
140917  November 6, 2012                            NaN           0  
140918  November 6, 2012                            NaN           0  
140919  November 6, 2012                            NaN           0  

[97500 rows x 6 columns]

In [25]:
# drop rows with 0 in "vote_count" column
tn_2012_namemelt=tn_2012_namemelt[tn_2012_namemelt['vote_count'] != 0].reset_index(drop=True)
tn_2012_namemelt

county       precinct                                   officename  \
0        ANDERSON  Andersonville                      United States President   
1        ANDERSON  Andersonville                         United States Senate   
2        ANDERSON  Andersonville       U.S. House of Representatives District   
3        ANDERSON  Andersonville  Tennessee House of Representatives District   
4        ANDERSON     Briceville                      United States President   
...           ...            ...                                          ...   
43415  WILLIAMSON           12-1                         United States Senate   
43416  WILLIAMSON           12-2                         United States Senate   
43417  WILLIAMSON           12-3                         United States Senate   
43418      WILSON           16-1                         United States Senate   
43419      WILSON           22-1                         United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012      . Chuck Fleischmann - (R)         996  
3      November 6, 2012          . Dennis Powers - (R)         998  
4      November 6, 2012            . Mitt Romney - (R)         227  
...                 ...                            ...         ...  
43415  November 6, 2012  . Write-In - Angelia Stinnett           2  
43416  November 6, 2012  . Write-In - Angelia Stinnett          13  
43417  November 6, 2012  . Write-In - Angelia Stinnett           3  
43418  November 6, 2012  . Write-In - Angelia Stinnett           1  
43419  November 6, 2012  . Write-In - Angelia Stinnett           1  

[43420 rows x 6 columns]

In [31]:
tn_2012_namemelt['vote_count'].value_counts()

1       4472
2       3531
3       2819
4       2166
5       1669
        ... 
1766       1
2611       1
1670       1
1638       1
4839       1
Name: vote_count, Length: 2260, dtype: int64

In [36]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2012_int=tn_2012_namemelt.groupby(['officename','county','candidate'], as_index=False)['vote_count'].sum()
tn_2012_int=tn_2012_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2012_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   officename  2445 non-null   object
 1   county      2445 non-null   object
 2   candidate   2445 non-null   object
 3   vote_count  2445 non-null   int64 
dtypes: int64(1), object(3)
memory usage: 76.5+ KB


In [37]:
tn_2012_int['officename'].unique()

array(['United States Senate', 'United States President',
       'U.S. House of Representatives District',
       'Tennessee House of Representatives District',
       'Tennessee Senate District'], dtype=object)

In [136]:
tn_2016_final['vote_count'].value_counts()

0       16824
1        2367
2        1670
3        1193
4         930
        ...  
1742        1
1710        1
1614        1
1582        1
2353        1
Name: vote_count, Length: 2211, dtype: int64

### 16824 rows have a vote count of 0. Drop those rows

In [137]:
tn_2016_final=tn_2016_final[tn_2016_final['vote_count'] != 0]
tn_2016_final=tn_2016_final.reset_index(drop=True)
tn_2016_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville  United States House of Representatives District 3   
2  Anderson  Andersonville     Tennessee House of Representatives District 36   
3  Anderson     Briceville                            United States President   
4  Anderson     Briceville  United States House of Representatives District 3   

    electdate               name       party  vote_count  
0  2016-11-08    Donald J. Trump  Republican        1222  
1  2016-11-08  Chuck Fleischmann  Republican        1124  
2  2016-11-08   Dennis H. Powers  Republican        1128  
3  2016-11-08    Donald J. Trump  Republican         243  
4  2016-11-08  Chuck Fleischmann  Republican         207

In [138]:
# confirm no 0's in vote count
tn_2016_final['vote_count'].value_counts()

1       2367
2       1670
3       1193
4        930
5        714
        ... 
1079       1
2998       1
2966       1
2934       1
2049       1
Name: vote_count, Length: 2210, dtype: int64

In [139]:
# create data frame tn_2016_pres for presidental election
tn_2016_pres_int=tn_2016_final[tn_2016_final['officename'].str.match('United States President')]
tn_2016_pres_int=tn_2016_pres_int.reset_index(drop=True)
tn_2016_pres_int

county       precinct               officename   electdate  \
0      Anderson  Andersonville  United States President  2016-11-08   
1      Anderson     Briceville  United States President  2016-11-08   
2      Anderson       Bull Run  United States President  2016-11-08   
3      Anderson        Clinton  United States President  2016-11-08   
4      Anderson   Clinton High  United States President  2016-11-08   
...         ...            ...                      ...         ...   
14085    Wilson            1-1  United States President  2016-11-08   
14086    Wilson           11-1  United States President  2016-11-08   
14087    Wilson           14-1  United States President  2016-11-08   
14088    Wilson           16-1  United States President  2016-11-08   
14089    Wilson           23-1  United States President  2016-11-08   

                          name        party  vote_count  
0              Donald J. Trump   Republican        1222  
1              Donald J. Trump   Republican         243  
2              Donald J. Trump   Republican        1397  
3              Donald J. Trump   Republican        1330  
4              Donald J. Trump   Republican         792  
...                        ...          ...         ...  
14085  Write-In - Tom Hoefling  Independent           1  
14086  Write-In - Tom Hoefling  Independent           2  
14087  Write-In - Tom Hoefling  Independent           1  
14088  Write-In - Tom Hoefling  Independent           1  
14089  Write-In - Tom Hoefling  Independent           1  

[14090 rows x 7 columns]

In [140]:
tn_2016_pres_int.head(20)

county        precinct               officename   electdate  \
0   Anderson   Andersonville  United States President  2016-11-08   
1   Anderson      Briceville  United States President  2016-11-08   
2   Anderson        Bull Run  United States President  2016-11-08   
3   Anderson         Clinton  United States President  2016-11-08   
4   Anderson    Clinton High  United States President  2016-11-08   
5   Anderson         Claxton  United States President  2016-11-08   
6   Anderson  Clinton Middle  United States President  2016-11-08   
7   Anderson    Dutch Valley  United States President  2016-11-08   
8   Anderson    Emory Valley  United States President  2016-11-08   
9   Anderson        Fairview  United States President  2016-11-08   
10  Anderson     Glen Alpine  United States President  2016-11-08   
11  Anderson        Glenwood  United States President  2016-11-08   
12  Anderson   Hendrix Creek  United States President  2016-11-08   
13  Anderson   Highland View  United States President  2016-11-08   
14  Anderson    Lake City MS  United States President  2016-11-08   
15  Anderson          Marlow  United States President  2016-11-08   
16  Anderson   North Clinton  United States President  2016-11-08   
17  Anderson          Norris  United States President  2016-11-08   
18  Anderson         Norwood  United States President  2016-11-08   
19  Anderson       Oak Ridge  United States President  2016-11-08   

               name       party  vote_count  
0   Donald J. Trump  Republican        1222  
1   Donald J. Trump  Republican         243  
2   Donald J. Trump  Republican        1397  
3   Donald J. Trump  Republican        1330  
4   Donald J. Trump  Republican         792  
5   Donald J. Trump  Republican        1319  
6   Donald J. Trump  Republican         453  
7   Donald J. Trump  Republican         271  
8   Donald J. Trump  Republican        1034  
9   Donald J. Trump  Republican         543  
10  Donald J. Trump  Republican        1149  
11  Donald J. Trump  Republican         763  
12  Donald J. Trump  Republican         849  
13  Donald J. Trump  Republican         299  
14  Donald J. Trump  Republican        1219  
15  Donald J. Trump  Republican         505  
16  Donald J. Trump  Republican         809  
17  Donald J. Trump  Republican         412  
18  Donald J. Trump  Republican         915  
19  Donald J. Trump  Republican         324

### Add column that has total vote count by count for use in calculating county vote percentages

In [141]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2016_pres_int=tn_2016_pres_int.groupby(['officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2016_pres=tn_2016_pres_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2016_pres.head(20)

officename      county                           name  \
0   United States President      Wilson                Donald J. Trump   
1   United States President      Wilson                Hillary Clinton   
2   United States President      Wilson                   Gary Johnson   
3   United States President      Wilson                     Jill Stein   
4   United States President      Wilson       Write-In - Evan McMullin   
5   United States President      Wilson                     Mike Smith   
6   United States President      Wilson                 Alyson Kennedy   
7   United States President      Wilson     "Rocky" Roque De La Fuente   
8   United States President      Wilson   Write-In - Darrell L. Castle   
9   United States President      Wilson        Write-In - Tom Hoefling   
10  United States President      Wilson  Write-In - Laurence Kotlikoff   
11  United States President  Williamson                Donald J. Trump   
12  United States President  Williamson                Hillary Clinton   
13  United States President  Williamson                   Gary Johnson   
14  United States President  Williamson       Write-In - Evan McMullin   
15  United States President  Williamson                     Jill Stein   
16  United States President  Williamson                     Mike Smith   
17  United States President  Williamson     "Rocky" Roque De La Fuente   
18  United States President  Williamson   Write-In - Darrell L. Castle   
19  United States President  Williamson                 Alyson Kennedy   

          party  vote_count  
0    Republican       39406  
1    Democratic       14385  
2   Independent        1807  
3   Independent         467  
4   Independent         280  
5   Independent         207  
6   Independent          83  
7   Independent          69  
8   Independent          23  
9   Independent           6  
10  Independent           1  
11   Republican       68212  
12   Democratic       31013  
13  Independent        4518  
14     No party        1371  
15  Independent         526  
16  Independent         318  
17  Independent         130  
18     No party         102  
19  Independent          67

In [142]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2016_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       39406
1      Wilson       14385
2      Wilson        1807
3      Wilson         467
4      Wilson         280
..        ...         ...
884  Anderson         174
885  Anderson         115
886  Anderson          74
887  Anderson          42
888  Anderson          28

[889 rows x 2 columns]

In [143]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       29881
Bedford        15355
Benton          6328
Bledsoe         4664
Blount         52208
               ...  
Wayne           5857
Weakley        12184
White           9825
Williamson    106271
Wilson         56734
Name: vote_count, Length: 95, dtype: int32

In [144]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2016_pres['sum_county_vote']=tn_2016_pres['county'].map(county_vote_sum)
tn_2016_pres.head(20)

officename      county                           name  \
0   United States President      Wilson                Donald J. Trump   
1   United States President      Wilson                Hillary Clinton   
2   United States President      Wilson                   Gary Johnson   
3   United States President      Wilson                     Jill Stein   
4   United States President      Wilson       Write-In - Evan McMullin   
5   United States President      Wilson                     Mike Smith   
6   United States President      Wilson                 Alyson Kennedy   
7   United States President      Wilson     "Rocky" Roque De La Fuente   
8   United States President      Wilson   Write-In - Darrell L. Castle   
9   United States President      Wilson        Write-In - Tom Hoefling   
10  United States President      Wilson  Write-In - Laurence Kotlikoff   
11  United States President  Williamson                Donald J. Trump   
12  United States President  Williamson                Hillary Clinton   
13  United States President  Williamson                   Gary Johnson   
14  United States President  Williamson       Write-In - Evan McMullin   
15  United States President  Williamson                     Jill Stein   
16  United States President  Williamson                     Mike Smith   
17  United States President  Williamson     "Rocky" Roque De La Fuente   
18  United States President  Williamson   Write-In - Darrell L. Castle   
19  United States President  Williamson                 Alyson Kennedy   

          party  vote_count  sum_county_vote  
0    Republican       39406            56734  
1    Democratic       14385            56734  
2   Independent        1807            56734  
3   Independent         467            56734  
4   Independent         280            56734  
5   Independent         207            56734  
6   Independent          83            56734  
7   Independent          69            56734  
8   Independent          23            56734  
9   Independent           6            56734  
10  Independent           1            56734  
11   Republican       68212           106271  
12   Democratic       31013           106271  
13  Independent        4518           106271  
14     No party        1371           106271  
15  Independent         526           106271  
16  Independent         318           106271  
17  Independent         130           106271  
18     No party         102           106271  
19  Independent          67           106271

In [145]:
# create "pct_county_vote" column for tn_2016_pres
tn_2016_pres['pct_county_vote']=round(tn_2016_pres['vote_count']/tn_2016_pres['sum_county_vote']*100,2)
tn_2016_pres

officename    county                          name  \
0    United States President    Wilson               Donald J. Trump   
1    United States President    Wilson               Hillary Clinton   
2    United States President    Wilson                  Gary Johnson   
3    United States President    Wilson                    Jill Stein   
4    United States President    Wilson      Write-In - Evan McMullin   
..                       ...       ...                           ...   
884  United States President  Anderson                    Mike Smith   
885  United States President  Anderson      Write-In - Evan McMullin   
886  United States President  Anderson    "Rocky" Roque De La Fuente   
887  United States President  Anderson                Alyson Kennedy   
888  United States President  Anderson  Write-In - Darrell L. Castle   

           party  vote_count  sum_county_vote  pct_county_vote  
0     Republican       39406            56734            69.46  
1     Democratic       14385            56734            25.36  
2    Independent        1807            56734             3.19  
3    Independent         467            56734             0.82  
4    Independent         280            56734             0.49  
..           ...         ...              ...              ...  
884  Independent         174            29881             0.58  
885     No party         115            29881             0.38  
886  Independent          74            29881             0.25  
887  Independent          42            29881             0.14  
888     No party          28            29881             0.09  

[889 rows x 7 columns]

In [146]:
# how many parties?
tn_2016_pres['party'].value_counts()

Independent    504
No party       195
Democratic      95
Republican      95
Name: party, dtype: int64

### Create dataframe "tn_2016_county_pres_pct"  with percent values per party per county

In [147]:
# create new dataframe with reduced columns values only
tn_2016_pct_county=tn_2016_pres.loc[0:,['county','party','pct_county_vote']]
tn_2016_pct_county

county        party  pct_county_vote
0      Wilson   Republican            69.46
1      Wilson   Democratic            25.36
2      Wilson  Independent             3.19
3      Wilson  Independent             0.82
4      Wilson  Independent             0.49
..        ...          ...              ...
884  Anderson  Independent             0.58
885  Anderson     No party             0.38
886  Anderson  Independent             0.25
887  Anderson  Independent             0.14
888  Anderson     No party             0.09

[889 rows x 3 columns]

In [148]:
# create new dataframe with Republican values only
tn_2016_rep_pct=tn_2016_pct_county[tn_2016_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2016_rep_pct

county       party  pct_county_vote
0       Wilson  Republican            69.46
1   Williamson  Republican            64.19
2        White  Republican            78.08
3      Weakley  Republican            73.93
4        Wayne  Republican            85.98
..         ...         ...              ...
90      Blount  Republican            71.72
91     Bledsoe  Republican            77.66
92      Benton  Republican            74.53
93     Bedford  Republican            74.80
94    Anderson  Republican            64.30

[95 rows x 3 columns]

In [149]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2016_rep_pct=tn_2016_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2016_rep_pct

county  Republican
0       Wilson       69.46
1   Williamson       64.19
2        White       78.08
3      Weakley       73.93
4        Wayne       85.98
..         ...         ...
90      Blount       71.72
91     Bledsoe       77.66
92      Benton       74.53
93     Bedford       74.80
94    Anderson       64.30

[95 rows x 2 columns]

In [150]:
# create new dataframe with Democratic values only
tn_2016_dem_pct=tn_2016_pct_county[tn_2016_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2016_dem_pct

county       party  pct_county_vote
0       Wilson  Democratic            25.36
1   Williamson  Democratic            29.18
2        White  Democratic            18.78
3      Weakley  Democratic            22.75
4        Wayne  Democratic            12.24
..         ...         ...              ...
90      Blount  Democratic            23.18
91     Bledsoe  Democratic            19.23
92      Benton  Democratic            23.29
93     Bedford  Democratic            22.11
94    Anderson  Democratic            30.16

[95 rows x 3 columns]

In [151]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2016_dem_pct=tn_2016_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2016_dem_pct

county  Democratic
0       Wilson       25.36
1   Williamson       29.18
2        White       18.78
3      Weakley       22.75
4        Wayne       12.24
..         ...         ...
90      Blount       23.18
91     Bledsoe       19.23
92      Benton       23.29
93     Bedford       22.11
94    Anderson       30.16

[95 rows x 2 columns]

In [152]:
# create new dataframe with non Republican or Democratic values only
tn_2016_other_pct=tn_2016_pct_county[(tn_2016_pct_county['party']=="Independent")|(tn_2016_pct_county['party']=="No party")].reset_index(drop=True)
tn_2016_other_pct

county        party  pct_county_vote
0      Wilson  Independent             3.19
1      Wilson  Independent             0.82
2      Wilson  Independent             0.49
3      Wilson  Independent             0.36
4      Wilson  Independent             0.15
..        ...          ...              ...
694  Anderson  Independent             0.58
695  Anderson     No party             0.38
696  Anderson  Independent             0.25
697  Anderson  Independent             0.14
698  Anderson     No party             0.09

[699 rows x 3 columns]

In [153]:
# sum by county
tn_2016_other_pct=tn_2016_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2016_other_pct

county  pct_county_vote
0     Anderson             5.53
1      Bedford             3.08
2       Benton             2.18
3      Bledsoe             3.11
4       Blount             5.10
..         ...              ...
90       Wayne             1.78
91     Weakley             3.32
92       White             3.14
93  Williamson             6.62
94      Wilson             5.18

[95 rows x 2 columns]

In [154]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2016_other_pct=tn_2016_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2016_other_pct

county  Other
0     Anderson   5.53
1      Bedford   3.08
2       Benton   2.18
3      Bledsoe   3.11
4       Blount   5.10
..         ...    ...
90       Wayne   1.78
91     Weakley   3.32
92       White   3.14
93  Williamson   6.62
94      Wilson   5.18

[95 rows x 2 columns]

In [155]:
# join tn_2016_rep, tn_2016_dem_pct and tn_2016_other_pct to form tn_2016_county_pres_pct
dem=tn_2016_dem_pct['Democratic']
other=tn_2016_other_pct['Other']
tn_2016_county_pres_pct=tn_2016_rep_pct.join(dem).join(other)
tn_2016_county_pres_pct

county  Republican  Democratic  Other
0       Wilson       69.46       25.36   5.53
1   Williamson       64.19       29.18   3.08
2        White       78.08       18.78   2.18
3      Weakley       73.93       22.75   3.11
4        Wayne       85.98       12.24   5.10
..         ...         ...         ...    ...
90      Blount       71.72       23.18   1.78
91     Bledsoe       77.66       19.23   3.32
92      Benton       74.53       23.29   3.14
93     Bedford       74.80       22.11   6.62
94    Anderson       64.30       30.16   5.18

[95 rows x 4 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [159]:
tn_2016_county_pres_pct['Rep_Margin']= tn_2016_county_pres_pct['Republican']-tn_2016_county_pres_pct['Democratic']
tn_2016_county_pres_pct.head(25)

county  Republican  Democratic  Other  Rep_Margin
0       Wilson       69.46       25.36   5.53       44.10
1   Williamson       64.19       29.18   3.08       35.01
2        White       78.08       18.78   2.18       59.30
3      Weakley       73.93       22.75   3.11       51.18
4        Wayne       85.98       12.24   5.10       73.74
5   Washington       68.85       26.18   4.72       42.67
6       Warren       70.09       25.97   3.30       44.12
7    Van Buren       75.24       22.28   3.39       52.96
8        Union       80.89       16.20   2.89       64.69
9       Unicoi       78.82       17.54   3.47       61.28
10   Trousdale       66.55       29.94   4.70       36.61
11      Tipton       72.02       24.64   2.81       47.38
12      Sumner       70.11       25.40   2.85       44.71
13    Sullivan       75.43       20.20   2.47       55.23
14     Stewart       72.92       23.06   2.93       49.86
15       Smith       73.88       22.71   4.08       51.17
16      Shelby       34.48       61.95   2.08      -27.47
17      Sevier       78.84       17.34   2.83       61.50
18  Sequatchie       77.99       18.49   6.29       59.50
19       Scott       84.85       13.11   3.09       71.74
20  Rutherford       60.05       34.17   2.39       25.88
21   Robertson       71.59       24.48   3.99       47.11
22       Roane       73.28       22.32   2.55       50.96
23        Rhea       78.29       17.56   2.40       60.73
24      Putnam       69.83       25.18   2.65       44.65

In [158]:
tn_2016_county_pres_pct.to_excel(r'../data/2016_election/tn_2016_county_pres_pct.xlsx',index=False)

In [108]:
#tn_2016_pct_county['rep_pct']=tn_2016_pct_county['county'].map(tn_2016_rep_pct.set_index('county')['rep_pct'])
#tn_2016_pct_county.head(25)

In [62]:
#tn_2016_pct_county['Rep-margin']=tn_2016_pct_county['rep_pct']-tn_2016_pct_county['pct_county_vote']
#tn_2016_pct_county

KeyError: 'rep_pct'

In [45]:
# Confirm negative numbers
#tn_2016_pct_county['repub_vic_margin'].unique()

array([  0.  ,  44.1 ,  66.27,  68.64,  68.97,  69.1 ,  69.31,  69.34,
        69.42,  69.45,  69.46,  35.01,  59.94,  62.9 ,  63.7 ,  63.89,
        64.07,  64.09,  64.13,  64.18,  64.19,  59.3 ,  76.56,  77.62,
        77.68,  77.82,  77.86,  78.04,  78.06,  51.18,  72.12,  73.42,
        73.58,  73.61,  73.72,  73.85,  73.89,  73.74,  85.02,  85.78,
        85.79,  85.83,  85.84,  85.89,  85.93,  42.67,  65.9 ,  67.89,
        68.43,  68.54,  68.69,  68.74,  68.78,  44.12,  67.6 ,  69.64,
        69.79,  69.83,  69.85,  69.95,  70.04,  70.08,  52.96,  73.34,
        74.91,  75.16,  75.2 ,  64.69,  79.16,  80.51,  80.59,  80.71,
        80.73,  80.76,  80.86,  61.28,  76.89,  78.08,  78.44,  78.58,
        78.63,  78.71,  78.76,  36.61,  64.24,  66.23,  66.3 ,  66.36,
        66.39,  66.42,  47.38,  71.64,  71.75,  71.79,  71.85,  71.88,
        71.93,  44.71,  67.17,  69.59,  69.6 ,  69.86,  69.97,  70.01,
        70.11,  55.23,  72.55,  74.86,  75.16,  75.28,  75.31,  75.34,
      

In [46]:
# join tn_2016_margin 'rep_vic_margin' column to tn_2016_pres
#rep_pct=tn_2016_pct_county['repub_vic_margin']
#tn_2016_pres=tn_2016_pres.join(rep_pct)
#tn_2016_pres.head(3)

officename  county             name        party  vote_count  \
0  United States President  Wilson  Donald J. Trump   Republican       39406   
1  United States President  Wilson  Hillary Clinton   Democratic       14385   
2  United States President  Wilson     Gary Johnson  Independent        1807   

   sum_county_vote  pct_county_vote  repub_vic_margin  
0            56734            69.46              0.00  
1            56734            25.36             44.10  
2            56734             3.19             66.27